From c78cd45b08561c79fe805ea63432be160a2f2e1c Mon Sep 17 00:00:00 2001
From: Marko Lindqvist <cazfi74@gmail.com>
Date: Wed, 15 Nov 2023 07:55:06 +0200
Subject: [PATCH 2/2] Pick random nations before setting turn number to 1

The consequences of having nations unassigned at turn 1
included failed asserts.

Reported by gatorized

https://forum.freeciv.org/f/viewtopic.php?t=94637

Signed-off-by: Marko Lindqvist <cazfi74@gmail.com>
---
 server/plrhand.c  |  2 +-
 server/srv_main.c | 11 ++++++-----
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/server/plrhand.c b/server/plrhand.c
index 98aff46cbf..76b91d346b 100644
--- a/server/plrhand.c
+++ b/server/plrhand.c
@@ -1385,7 +1385,7 @@ static void package_player_info(struct player *plr,
       packet->color_green = preferred->g;
       packet->color_blue = preferred->b;
     } else {
-      fc_assert(game.info.turn < 1);
+      fc_assert(game.info.turn < 1); /* Game has not yet started */
       packet->color_valid = FALSE;
       /* Client shouldn't use these dummy values */
       packet->color_red = 0;
diff --git a/server/srv_main.c b/server/srv_main.c
index a704497d03..9eae8cc5ae 100644
--- a/server/srv_main.c
+++ b/server/srv_main.c
@@ -2317,7 +2317,7 @@ void update_nations_with_startpos(void)
 }
 
 /**********************************************************************//**
-  Handles a pick-nation packet from the client.  These packets are
+  Handles a pick-nation packet from the client. These packets are
   handled by connection because ctrl users may edit anyone's nation in
   pregame, and editing is possible during a running game.
 **************************************************************************/
@@ -3199,16 +3199,17 @@ static void srv_ready(void)
 #endif
 
   if (game.info.is_new_game) {
-    game.info.turn++; /* pregame T0 -> game T1 */
-    fc_assert(game.info.turn == 1);
-    game.info.year = game.server.start_year;
     /* Must come before assign_player_colors() */
     generate_players();
     final_ruleset_adjustments();
+
+    game.info.turn++; /* Pregame T0 -> game T1 */
+    fc_assert(game.info.turn == 1);
+    game.info.year = game.server.start_year;
   }
 
   /* If we have a tile map, and MAPGEN_SCENARIO == map.server.generator,
-   * call map_fractal_generate anyway to make the specials, huts and
+   * call map_fractal_generate() anyway to make the specials, huts and
    * continent numbers. */
   if (map_is_empty()
       || (MAPGEN_SCENARIO == wld.map.server.generator
-- 
2.42.0

